include/os/linux/zfs/sys/zfs_znode_impl.h

```diff
@@ -82,18 +82,33 @@ extern "C" {
 #define	zhold(zp)	VERIFY3P(igrab(ZTOI((zp))), !=, NULL)
 #define	zrele(zp)	iput(ZTOI((zp)))
 
+#define	zfsvfs_is_unmounted(zfsvfs)				\
+	((zfsvfs)->z_unmounted && (zfsvfs)->z_force_unmounted)
+
 /* Called on entry to each ZFS inode and vfs operation. */
 #define	ZFS_ENTER_ERROR(zfsvfs, error)				\
 do {								\
 	ZFS_TEARDOWN_ENTER_READ(zfsvfs, FTAG);			\
-	if (unlikely((zfsvfs)->z_unmounted)) {			\
+	if (unlikely(zfsvfs_is_unmounted(zfsvfs))) {			\
 		ZFS_TEARDOWN_EXIT_READ(zfsvfs, FTAG);		\
 		return (error);					\
 	}							\
 } while (0)
 #define	ZFS_ENTER(zfsvfs)	ZFS_ENTER_ERROR(zfsvfs, EIO)
 #define	ZPL_ENTER(zfsvfs)	ZFS_ENTER_ERROR(zfsvfs, -EIO)
 
+/* ZFS_ENTER but ok with forced unmount having begun */
+#define	_ZFS_ENTER_UNMOUNTOK(zfsvfs, error)			\
+do {								\
+	rrm_enter_read(&(zfsvfs)->z_teardown_lock, FTAG);	\
+	if ((zfsvfs)->z_unmounted == B_TRUE) {			\
+		ZFS_EXIT(zfsvfs);				\
```

Comment:
I think this is already addressed.

> call it ZFS_ENTER_ERROR_UNMOUNTOK
In 768eacedef we have changed the naming of the corresponding functions. I think currently the name is adequate for the others.

> call ZFS_TEARDOWN_ENTER_READ() instead of rrm_enter_read()
Done.

> call ZFS_TEARDOWN_EXIT_READ() instead of ZFS_EXIT()
We are calling zfs_exit, as zfs_exit does more then ZFS_TEARDOWN_EXIT_READ (it also calls `zfs_exit_fs`).